pkg/lib/long-running-process.js

```diff
@@ -0,0 +1,126 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Manage a long-running precious process that runs independently from a Cockpit
+ * session in a transient systemd service unit. See
+ * examples/long-running-process/README.md for details. */
+
+/* global cockpit */
+
+// systemd D-Bus API names
+const O_SD_OBJ = "/org/freedesktop/systemd1";
+const I_SD_MGR = "org.freedesktop.systemd1.Manager";
+const I_SD_UNIT = "org.freedesktop.systemd1.Unit";
+const I_DBUS_PROP = "org.freedesktop.DBus.Properties";
+
+export default class LongRunningProcess {
+    /* serviceName: systemd unit name to start or reattach to
+     * updateCallback: function that gets called whenever the state changed; first and only
+     *                 argument is `this` LongRunningProcess instance.
+     */
+    constructor(serviceName, updateCallback) {
+        // don't require superuser; this is only for reading the current state
+        this.systemdClient = cockpit.dbus("org.freedesktop.systemd1");
+        this.serviceName = serviceName;
+        this.updateCallback = updateCallback;
+        this.setRunning(null);
+        this.startTimestamp = null; // µs since epoch
+
+        // Watch for start event of the service
+        this.systemdClient.subscribe({ interface: I_SD_MGR, member: "JobNew" }, (path, iface, signal, args) => {
+            if (args[2] == this.serviceName)
+                this.checkState();
+        });
+
+        // Check if it is already running
+        this.checkState();
+    }
+
+    // null: initializing; false: not running; true: running, "failed": unit failed
+    setRunning(running) {
+        /* PropertiesChanged often gets fired multiple times with the same values, avoid UI flicker */
+        if (running === this.running)
+            return;
+        this.running = running;
+        if (this.updateCallback)
+            this.updateCallback(this);
+    }
+
+    setRunningFromProperties(activeState, stateChangeTimestamp) {
+        if (activeState === 'activating') {
+            this.startTimestamp = stateChangeTimestamp;
+            this.setRunning(true);
+        } else if (activeState === 'failed') {
+            this.startTimestamp = null; // TODO: can we derive this from InvocationID?
+            this.setRunning("failed");
```

Comment:
This lays a trap for code like `if (process.running)`, which would evaluate to true with `running == "failed"`.

What about a general `process.state`, maybe in addition to a strictly boolean `process.running?`

We could have `null`, `"running"`, `"failed"`, and `"terminated"` values for `state`.

Reply:
I'll think about that. Possibly we also just directly plumb through the `ActiveState` systemd property? At first I wanted to simplify this a little bit and detach it from the underlying implementation. Having *both* process.state and process.running seems even more confusing IMHO, as the caller is prone to just check .running without .state?

Reply:
> as the caller is prone to just check .running without .state?

But that would be fine, no?

Reply:
No, it would not be fine, as it opens up a race condition. During initialization, `process.running` must have some value, and neither "true" nor "false" is right -- this could lead the consumer into trying to start a process (because `!running)` even thouguh it is already running. I don't see how you could only ever look at a boolean "running", you *at least* need a tristate.

I'm fine with cleaning up the API so that `process.state` is always a string, including "init" instead of `null`. Then at least it's obvious that it's never false-y.